packages/ember-metal/lib/properties.js

```diff
@@ -83,61 +83,71 @@ export function DEFAULT_GETTER_FUNCTION(name) {
-  var possibleDesc, existingDesc, watching, value;
+export function defineProperty(obj, keyName, desc, data/*, meta*/) {
+  let meta = arguments[4] || metaFor(obj);
+  let watchEntry = meta.peekWatching(keyName);
+  let existingDesc = meta.peekDescs(keyName);
 
-  if (!meta) {
-    meta = metaFor(obj);
-  }
-  var watchEntry = meta.peekWatching(keyName);
-  possibleDesc = obj[keyName];
-  existingDesc = (possibleDesc !== null && typeof possibleDesc === 'object' && possibleDesc.isDescriptor) ? possibleDesc : undefined;
-
-  watching = watchEntry !== undefined && watchEntry > 0;
+  let watching = watchEntry !== undefined && watchEntry > 0;
 
   if (existingDesc) {
     existingDesc.teardown(obj, keyName);
   }
 
+  let value;
+
+  // is this instanceof needed?
   if (desc instanceof Descriptor) {
-    value = desc;
-    if (isEnabled('mandatory-setter')) {
-      if (watching) {
-        Object.defineProperty(obj, keyName, {
-          configurable: true,
-          enumerable: true,
-          writable: true,
-          value: value
-        });
-      } else {
+    meta.writeDescs(keyName, desc);
+
+    Object.defineProperty(obj, keyName, {
+      configurable: true,
+      enumerable: true,
+      get() {
+        return metaFor(this).peekDescs(keyName).get(this, keyName);
+      },
+      set(value) {
+        // throw TypeError('Sorry this isnt supported "right now"');
+        delete obj[keyName];
         obj[keyName] = value;
       }
-    } else {
-      obj[keyName] = value;
-    }
+    });
+
+    // if (isEnabled('mandatory-setter')) {
+    //   if (watching) {
+    //     Object.defineProperty(obj, keyName, {
+    //       configurable: true,
+    //       enumerable: true,
+    //       writable: true,
+    //       value: value
+    //     });
+    //   } else {
+    //     obj[keyName] = value;
+    //   }
+    // } else {
+    //   obj[keyName] = value;
+    // }
     if (desc.setup) { desc.setup(obj, keyName); }
   } else {
+    // add insert undefined subroutine, to do this more efficiently
+    meta.writeDescs(keyName, false);
     if (desc == null) {
-      value = data;
-
-      if (isEnabled('mandatory-setter')) {
-        if (watching) {
-          meta.writeValues(keyName, data);
-          Object.defineProperty(obj, keyName, {
-            configurable: true,
-            enumerable: true,
-            set: MANDATORY_SETTER_FUNCTION(keyName),
-            get: DEFAULT_GETTER_FUNCTION(keyName)
-          });
-        } else {
-          obj[keyName] = data;
-        }
+      meta.writeValues(keyName, data);
+      if (isEnabled('mandatory-setter') && watching) {
+        Object.defineProperty(obj, keyName, {
+          configurable: true,
+          enumerable: true,
+          set: MANDATORY_SETTER_FUNCTION(keyName),
+          get: DEFAULT_GETTER_FUNCTION(keyName)
+        });
       } else {
-        obj[keyName] = data;
+        Object.defineProperty(obj, keyName, {
```

Comment:
Would be nice if we could avoid this, but we do stop on CP's sometimes, which requires us to remove the existing descriptor. A likely approach is to check if this property has a desc or not, then decide defProp or not.